### PR TITLE
Adding unittests to test the collect_and_send_events call

### PR DESCRIPTION
**Pull request by vrdmr**
## Description
The collect_and_send_events code needed unittests to verify its workings.

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1564)

### CODE REVIEW

File: azurelinuxagent/common/protocol/wire.py [hunk 1]
```diff
@@ -1137,7 +1137,9 @@ def report_event(self, event_list):
                 buf[event.providerId] = ""
             event_str = event_to_v1(event)
             if len(event_str) >= 63 * 1024:
-                logger.warn("Single event too large: {0}", event_str[300:])
+                details_of_event = [ustr(x.name) + ":" + ustr(x.value) for x in event.parameters if x.name in
+                                    ["Name", "Version", "Operation", "OperationSuccess"]]
+                logger.warn("Single event too large: {0}".format(str(details_of_event)))
```
Comment (vrdmr) [inline azurelinuxagent/common/protocol/wire.py]: Main change - fixing the log for large size messages.

File: azurelinuxagent/ga/monitor.py [hunk 2]
```diff
@@ -176,7 +176,7 @@ def init_sysinfo(self):
             self.sysinfo.append(TelemetryEventParam("RAM", ram))
```
Comment (vrdmr) [inline azurelinuxagent/ga/monitor.py]: This file mainly changes for correctly formatting exceptions.

File: tests/ga/test_monitor.py [hunk 3]
```diff
@@ -229,6 +277,153 @@ def test_failed_heartbeat_creates_telemetry(self, patch_report_heartbeat, _, *ar
...
+
+        for power in sizes:
+            size = 2 ** power * 1024
+            self.event_logger.save_event(create_event_message(size))
+
+        with patch("azurelinuxagent.common.utils.restutil.http_post") as mock_http_post:
+            mock_http_post.return_value = ResponseMock(
+                status=restutil.httpclient.SERVICE_UNAVAILABLE,
+                response="")
+            monitor_handler.collect_and_send_events()
```
Comment (pgombar) [inline tests/ga/test_monitor.py]: Missing an assert here?

File: tests/ga/test_monitor.py [hunk 4]
```diff
@@ -229,6 +277,153 @@ def test_failed_heartbeat_creates_telemetry(self, patch_report_heartbeat, _, *ar
...
+
+        sizes = [1, 2, 3]  # get the powers of 2, and multiple by 1024.
+
+        for power in sizes:
+            size = 2 ** power * 1024
+            self.event_logger.save_event(create_event_message(size))
+
+        with patch("azurelinuxagent.common.protocol.wire.WireClient.call_wireserver") as patch_call_wireserver:
+            patch_call_wireserver.side_effect = HttpError
+            monitor_handler.collect_and_send_events()
```
Comment (pgombar) [inline tests/ga/test_monitor.py]: Potentially also missing an assert here.

File: tests/utils/test_process_util.py [hunk 5]
```diff
@@ -92,3 +92,21 @@ def test_format_stdout_stderr05(self):
         expected = '[stdout]\n\n\n[stderr]\n'
         actual = format_stdout_stderr('', '', 1000)
         self.assertEqual(expected, actual)
+
+    def test_read_output_valid_stdout_stderr(self):
```
Comment (pgombar) [inline tests/utils/test_process_util.py]: Unit tests for this method are already checked in develop: https://github.com/Azure/WALinuxAgent/blob/develop/tests/utils/test_process_util.py#L22

Comment (vrdmr) [reply tests/utils/test_process_util.py]: I'll remove this test and put the tests in a new PR which would also test non-ascii text.

File: azurelinuxagent/common/protocol/wire.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: we need a corresponding unit test in tests/protocol/test_wire.py

Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: Let's add the limit we are checking against to the warning to facilitate debugging

File: azurelinuxagent/ga/monitor.py [hunk 6]
```diff
@@ -176,7 +176,7 @@ def init_sysinfo(self):
...
             self.sysinfo.append(TelemetryEventParam("Processors", processors))
         except OSUtilError as e:
-            logger.warn("Failed to get system info: {0}", e)
+            logger.warn("Failed to get system info: {0}".format(ustr(e)))
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: isn't logger.warn supposed to do the call to format()?

File: azurelinuxagent/ga/monitor.py [hunk 7]
```diff
@@ -237,15 +238,15 @@ def collect_and_send_events(self):
                     try:
                         data_str = self.collect_event(event_file_path)
                     except EventError as e:
-                        logger.error("{0}", e)
+                        logger.error(ustr(e))
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: >  logger.error("{0}", ustr(e))

in case ustr(e) includes braces

File: azurelinuxagent/ga/monitor.py [hunk 8]
```diff
@@ -254,9 +255,9 @@ def collect_and_send_events(self):
                 try:
                     self.protocol.report_event(event_list)
                 except ProtocolError as e:
-                    logger.error("{0}", e)
+                    logger.error(ustr(e))
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: logger.error("{0}", ustr(e))

**Comment by narrieta** [comment]
@vrdmr - how did we address reducing the frequency of those messages to the log (since it is echoed to the serial console)?

File: tests/common/test_event.py [hunk 9]
```diff
@@ -163,6 +165,23 @@ def test_save_event(self):
         add_event('test', message='test event')
         self.assertTrue(len(os.listdir(self.tmp_dir)) == 2)
 
+    def test_save_event_message_with_non_ascii_characters(self):
+        stdout = open(os.path.join(data_dir, "events", "collect_and_send_extension_stdout_stderror",
+                                   "dummy_stdout_with_non_ascii_characters"), mode="r+b")
+        stderr = open(os.path.join(data_dir, "events", "collect_and_send_extension_stdout_stderror",
```
Comment (narrieta) [inline tests/common/test_event.py]: need to close those 2 files

File: tests/common/test_event.py [hunk 10]
```diff
@@ -163,6 +165,23 @@ def test_save_event(self):
...
+        msg = read_output(stdout, stderr)
+        duration = elapsed_milliseconds(datetime.utcnow())
+        log_msg = "{0}\n{1}".format("DummyCmd", "\n".join([line for line in msg.split('\n') if line != ""]))
+
+        add_event('test_extension', message=log_msg, duration=duration)
+        for tld_file in os.listdir(self.tmp_dir):
+            event_str = MonitorHandler.collect_event(os.path.join(self.tmp_dir, tld_file))
+            event_json = json.loads(event_str)
+
+            self.assertEqual(8, len(event_json["parameters"]))
```
Comment (narrieta) [inline tests/common/test_event.py]: can we have a stronger assertion? e.g. all the expected parameters (Name, etc) are present?

Comment (narrieta) [inline tests/common/test_event.py]: can we add a stronger assert? e.g. all parameters (Name, etc) are present

File: tests/common/test_event.py [hunk 11]
```diff
@@ -163,6 +165,23 @@ def test_save_event(self):
...
+        stderr = open(os.path.join(data_dir, "events", "collect_and_send_extension_stdout_stderror",
+                                   "dummy_stderr_with_non_ascii_characters"), mode="r+b")
+
+        msg = read_output(stdout, stderr)
+        duration = elapsed_milliseconds(datetime.utcnow())
+        log_msg = "{0}\n{1}".format("DummyCmd", "\n".join([line for line in msg.split('\n') if line != ""]))
+
+        add_event('test_extension', message=log_msg, duration=duration)
+        for tld_file in os.listdir(self.tmp_dir):
+            event_str = MonitorHandler.collect_event(os.path.join(self.tmp_dir, tld_file))
```
Comment (narrieta) [inline tests/common/test_event.py]: can you fix the implementation of collect_event to cleanup (delete) the file if it cannot decode it?

```
 def collect_event(evt_file_name):
        try:
            logger.verbose("Found event file: {0}".format(evt_file_name))
            with open(evt_file_name, "rb") as evt_file:
                # if fail to open or delete the file, throw exception
                data_str = evt_file.read().decode("utf-8", 'ignore')
            logger.verbose("Processed event file: {0}".format(evt_file_name))
            os.remove(evt_file_name)
            return data_str
        except IOError as e:
            msg = "Failed to process {0}, {1}".format(evt_file_name, e)
            raise EventError(msg)
```

File: tests/ga/test_monitor.py [hunk 12]
```diff
@@ -229,6 +277,158 @@ def test_failed_heartbeat_creates_telemetry(self, patch_report_heartbeat, _, *ar
...
+        sizes = [1, 2, 3, 4, 5]  # get the powers of 2, and multiple by 1024.
+
+        for power in sizes:
+            size = 2 ** power * 1024
+            self.event_logger.save_event(create_event_message(size))
+        monitor_handler.collect_and_send_events()
+
+        # The send_event call would be called twice - Once when the buffer fills up from 1 to 4, and then for 5.
+
+        self.assertEqual(2, patch_send_event.call_count)
```
Comment (narrieta) [inline tests/ga/test_monitor.py]: can we add asserts verifying the arguments  to these 2 calls?

> Once when the buffer fills up from 1 to 4, and then for 5.

File: tests/ga/test_monitor.py [hunk 13]
```diff
@@ -229,6 +277,158 @@ def test_failed_heartbeat_creates_telemetry(self, patch_report_heartbeat, _, *ar
...
+        monitor_handler.init_protocols()
+
+        sizes = [9, 10, 11]  # get the powers of 2, and multiple by 1024.
+
+        for power in sizes:
+            size = 2 ** power * 1024
+            self.event_logger.save_event(create_event_message(size))
+        monitor_handler.collect_and_send_events()
+
+        # The send_event call should never be called as the events are larger than 2**16.
```
Comment (narrieta) [inline tests/ga/test_monitor.py]: can we add an assert on warning emitted on this case?

File: tests/ga/test_monitor.py [hunk 14]
```diff
@@ -229,6 +277,158 @@ def test_failed_heartbeat_creates_telemetry(self, patch_report_heartbeat, _, *ar
...
+            size = 2 ** power * 1024
+            self.event_logger.save_event(create_event_message(size))
+
+        with patch("azurelinuxagent.common.utils.restutil.http_post") as mock_http_post:
+            mock_http_post.return_value = ResponseMock(
+                status=restutil.httpclient.SERVICE_UNAVAILABLE,
+                response="")
+            monitor_handler.collect_and_send_events()
+
+        # This test validates that if we hit an issue while sending an event, we never send it again.
```
Comment (narrieta) [inline tests/ga/test_monitor.py]: do we emit a warning in that case? if yes, can we add an assert?

Comment (narrieta) [inline tests/ga/test_monitor.py]: > if we hit an issue while sending an event, we never send it again.

so a transitory issue may prevent us from sending telemetry? how hard would it be to improve that?

File: azurelinuxagent/ga/monitor.py (hunk 6 above)
Comment (vrdmr) [reply azurelinuxagent/ga/monitor.py]: Yes. Reverted the change (adding the ustr)

File: tests/ga/test_monitor.py (hunk 14 above)
Comment (vrdmr) [reply tests/ga/test_monitor.py]: We don't emit any warning in all the cases, but in some specific cases. I've broken the test into 3 distinct tests and also checking the events folder to see if we preserve on the disk or not.

Comment (vrdmr) [reply tests/ga/test_monitor.py]: Had an offline discussion. A new bug is created to investigate this issue.

File: tests/ga/test_monitor.py (hunk 12 above)
Comment (vrdmr) [reply tests/ga/test_monitor.py]: Changed the test to be more deterministic.

**Comment by vrdmr** [comment]
> @vrdmr - how did we address reducing the frequency of those messages to the log (since it is echoed to the serial console)?

I have changed the warning to use `periodic_warn()` and it is reset every 30 minutes. Thus we are reducing both the content written on the console as well as the count of the logs written.

File: azurelinuxagent/ga/monitor.py [hunk 15]
```diff
@@ -206,18 +206,20 @@ def init_sysinfo(self):
...
         except (HttpError, ValueError) as e:
-            logger.warn("failed to get IMDS info: {0}", e)
+            logger.warn("failed to get IMDS info: {0}", ustr(e))
 
-    def collect_event(self, evt_file_name):
+    @staticmethod
+    def collect_event(evt_file_name):
         try:
-            logger.verbose("Found event file: {0}", evt_file_name)
+            logger.verbose("Found event file: {0}".format(evt_file_name))
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: similarly to logger.warn, logger.verbose does the call to format()

File: azurelinuxagent/ga/monitor.py [hunk 16]
```diff
@@ -206,18 +206,20 @@ def init_sysinfo(self):
...
             with open(evt_file_name, "rb") as evt_file:
                 # if fail to open or delete the file, throw exception
                 data_str = evt_file.read().decode("utf-8", 'ignore')
-            logger.verbose("Processed event file: {0}", evt_file_name)
+            logger.verbose("Processed event file: {0}".format(evt_file_name))
             os.remove(evt_file_name)
             return data_str
         except IOError as e:
+            os.remove(evt_file_name)
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: why remove only on IOError? will decode() raise this exception?

Comment (vrdmr) [reply azurelinuxagent/ga/monitor.py]: We never get any error as we use `ignore` param in decode - `decode("utf-8", 'ignore')`.

I have still added another exception `UnicodeDecodeError` as a precaution. We can decide to remove the `ignore` flag.

Changing the `except IOError` to `except (IOError, UnicodeDecodeError)`

**Pull request merged**